### PR TITLE
[RFC] vim-patch:7.4.582

### PR DESCRIPTION
**Pull request by glts**
```
Problem:    Can't match "%>80v" properly. (Axel Bender)
Solution:   Correctly handle ">". (Christian Brabandt)
```

Repro https://groups.google.com/d/msg/vim_dev/n-02i4FnOcw/P3Yyx1OLeXgJ

https://github.com/vim/vim/commit/v7-4-582

Original patch:

``` diff
diff --git a/src/nvim/regexp_nfa.c b/src/nvim/regexp_nfa.c
--- a/src/nvim/regexp_nfa.c
+++ b/src/nvim/regexp_nfa.c
@@ -6441,16 +6441,26 @@ nfa_regmatch(prog, start, submatch, m)
        {
            int     op = t->state->c - NFA_VCOL;
            colnr_T col = (colnr_T)(reginput - regline);
+           win_T   *wp = reg_win == NULL ? curwin : reg_win;

            /* Bail out quickly when there can't be a match, avoid the
             * overhead of win_linetabsize() on long lines. */
-           if ((col > t->state->val && op != 1)
-               || (col - 1 > t->state->val && op == 1))
+           if (op != 1 && col > t->state->val)
            break;
-           result = nfa_re_num_cmp(t->state->val, op,
-           (long_u)win_linetabsize(
-               reg_win == NULL ? curwin : reg_win,
-                              regline, col) + 1);
+           result = FALSE;
+           if (op == 1 && col - 1 > t->state->val && col > 100)
+           {
+           int ts = wp->w_buffer->b_p_ts;
+
+           /* Guess that a character won't use more columns than
+            * 'tabstop', with a minimum of 4. */
+           if (ts < 4)
+               ts = 4;
+           result = col > t->state->val * ts;
+           }
+           if (!result)
+           result = nfa_re_num_cmp(t->state->val, op,
+               (long_u)win_linetabsize(wp, regline, col) + 1);
            if (result)
            {
            add_here = TRUE;
diff --git a/src/nvim/testdir/test64.in b/src/nvim/testdir/test64.in
--- a/src/nvim/testdir/test64.in
+++ b/src/nvim/testdir/test64.in
@@ -7,6 +7,7 @@ actually tried.
 STARTTEST
 :so small.vim
 :" tl is a List of Lists with:
+:"    regexp engine
 :"    regexp pattern
 :"    text to test the pattern on
 :"    expected match (optional)
@@ -451,6 +452,9 @@ STARTTEST
 :"""" Skip adding state twice
 :call add(tl, [2, '^\%(\%(^\s*#\s*if\>\|#\s*if\)\)\(\%>1c.*$\)\@=', "#if FOO", "#if", ' FOO'])
 :"
+:""" Test \%V atom
+:call add(tl, [2, '\%>70vGesamt', 'Jean-Michel Charlier & Victor Hubinon\Gesamtausgabe [Salleck]    Buck Danny {Jean-Michel Charlier & Victor Hubinon}\Gesamtausgabe', 'Gesamt'])
+:"
 :"""" Run the tests
 :"
 :for t in tl
diff --git a/src/nvim/testdir/test64.ok b/src/nvim/testdir/test64.ok
--- a/src/nvim/testdir/test64.ok
+++ b/src/nvim/testdir/test64.ok
@@ -1030,6 +1030,9 @@ OK 2 - [0-9a-zA-Z]\{8}-\([0-9a-zA-Z]\{4}
 OK 0 - ^\%(\%(^\s*#\s*if\>\|#\s*if\)\)\(\%>1c.*$\)\@=
 OK 1 - ^\%(\%(^\s*#\s*if\>\|#\s*if\)\)\(\%>1c.*$\)\@=
 OK 2 - ^\%(\%(^\s*#\s*if\>\|#\s*if\)\)\(\%>1c.*$\)\@=
+OK 0 - \%>70vGesamt
+OK 1 - \%>70vGesamt
+OK 2 - \%>70vGesamt
 multi-line tests
 OK 0 - ^.\(.\).\_..\1.
 OK 1 - ^.\(.\).\_..\1.
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    582,
+/**/
     581,
 /**/
     580,
```

### CODE REVIEW

**Comment by Shougo** [comment]
:+1: 

This is very important change for plugins.

File: src/nvim/regexp_nfa.c [hunk 1]
```diff
@@ -5742,17 +5742,28 @@ static int nfa_regmatch(nfa_regprog_T *prog, nfa_state_T *start, regsubs_T *subm
...
-                                          col);
-          assert(t->state->val >= 0);
-          result = nfa_re_num_cmp((uintmax_t)t->state->val,
-                                  op,
-                                  lts + 1);
+
+          result = FALSE;
+          win_T *wp = reg_win == NULL ? curwin : reg_win;
+          if (op == 1 && col - 1 > t->state->val && col > 100) {
+            assert(wp->w_buffer->b_p_ts <= INT_MAX);
```
Comment (fwalch) [inline src/nvim/regexp_nfa.c]: :+1:

**Comment by fwalch** [comment]
One small comment, otherwise LGTM.

File: src/nvim/regexp_nfa.c [hunk 2]
```diff
@@ -5742,17 +5742,28 @@ static int nfa_regmatch(nfa_regprog_T *prog, nfa_state_T *start, regsubs_T *subm
...
+            int ts = (int)wp->w_buffer->b_p_ts;
+
+            // Guess that a character won't use more columns than 'tabstop',
+            // with a minimum of 4.
+            if (ts < 4) {
+              ts = 4;
+            }
+            result = col > t->state->val * ts;
```
Comment (glts) [inline src/nvim/regexp_nfa.c]: > Hm, on second thought, couldn't you also just create `ts` as `long ts`?

It does seem so, yes. Just to be sure: making `ts` a `long` doesn't create a problem here, does it? the comparison will be done in `long`, won't it?

File: src/nvim/regexp_nfa.c (hunk 1 above)
Comment (glts) [inline src/nvim/regexp_nfa.c]: (My reply is a few lines below this.)

File: src/nvim/regexp_nfa.c (hunk 2 above)
Comment (fwalch) [inline src/nvim/regexp_nfa.c]: > the comparison will be done in long, won't it?

I believe it will, yes. So I think it should not be a problem.

**Comment by glts** [comment]
Rebased with only the change suggested in https://github.com/neovim/neovim/pull/2653#discussion_r30952975

```patch
diff --git a/src/nvim/regexp_nfa.c b/src/nvim/regexp_nfa.c
index 974d3c9..4ef9181 100644
--- a/src/nvim/regexp_nfa.c
+++ b/src/nvim/regexp_nfa.c
@@ -5749,8 +5749,7 @@ static int nfa_regmatch(nfa_regprog_T *prog, nfa_state_T *start, regsubs_T *subm
           result = FALSE;
           win_T *wp = reg_win == NULL ? curwin : reg_win;
           if (op == 1 && col - 1 > t->state->val && col > 100) {
-            assert(wp->w_buffer->b_p_ts <= INT_MAX);
-            int ts = (int)wp->w_buffer->b_p_ts;
+            long ts = wp->w_buffer->b_p_ts;
 
             // Guess that a character won't use more columns than 'tabstop',
             // with a minimum of 4.
```

**Comment by justinmk** [comment]
Merged, thanks! This fixed an important bug.

**Pull request closed without merging**